openedx/core/djangoapps/waffle_utils/__init__.py

```diff
@@ -1,503 +1,65 @@
-    _class_instances = WeakSet()
-
-    def __init__(self, waffle_namespace, switch_name, module_name=None):
-        """
-        Arguments:
-            waffle_namespace (WaffleSwitchNamespace | String): Namespace for this switch.
-            switch_name (String): The name of the switch (without namespacing).
-            module_name (String): The name of the module where the flag is created. This should be ``__name__`` in most
-            cases.
-        """
-        if isinstance(waffle_namespace, six.string_types):
-            waffle_namespace = WaffleSwitchNamespace(name=waffle_namespace)
-
-        self.waffle_namespace = waffle_namespace
-        self.switch_name = switch_name
-        self._module_name = module_name or ""
-        self._class_instances.add(self)
-
-    @classmethod
-    def get_instances(cls):
-        """ Returns a WeakSet of the instantiated instances of WaffleFlag. """
-        return cls._class_instances
+    Represents a single waffle flag that can be forced on/off for a course. This class should be used instead of
+    WaffleFlag when in the context of a course.
 
-    @property
-    def module_name(self):
-        """
-        Returns the module name. This is cached to work with the WeakSet instances.
-        """
-        return self._module_name
-
-    @module_name.setter
-    def module_name(self, value):
-        self._module_name = value
-
-    @property
-    def namespaced_switch_name(self):
-        """
-        Returns the fully namespaced switch name.
-        """
-        return self.waffle_namespace._namespaced_name(self.switch_name)  # pylint: disable=protected-access
+    Uses a cached waffle namespace.
 
-    def is_enabled(self):
-        return self.waffle_namespace.is_enabled(self.switch_name)
+    Usage:
 
-    @contextmanager
-    def override(self, active=True):
-        with self.waffle_namespace.override(self.switch_name, active):
-            yield
+       WAFFLE_FLAG_NAMESPACE = WaffleFlagNamespace(name='my_namespace')
+       SOME_COURSE_FLAG = CourseWaffleFlag(WAFFLE_FLAG_NAMESPACE, 'some_course_feature', __name__)
 
+    And then we can check this flag in code with::
 
-class WaffleFlagNamespace(six.with_metaclass(ABCMeta, WaffleNamespace)):
-    """
-    Provides a single namespace for a set of waffle flags.
+        SOME_COURSE_FLAG.is_enabled(course_key)
 
-    All namespaced flag values are stored in a single request cache containing
-    all flags for all namespaces.
+    The Django Admin "waffle_utils" section can be used to configure a course override for this same flag (e.g.
+    my_namespace.some_course_feature).
     """
 
-    @property
-    def _cached_flags(self):
-        """
-        Returns a dictionary of all namespaced flags in the request cache.
-        """
-        return self._get_request_cache().setdefault('flags', {})
-
-    def is_flag_active(self, flag_name, check_before_waffle_callback=None):
+    def _get_course_override_value(self, course_key):
```

Comment:
Nit: A better name would be `_get_and_cache_course_override_value`

Reply:
FYI: I got confused. I thought this was already in another branch, but didn't get what was going on.  This would have been simpler to review if this PR compared to the open branch (for now) instead of master, so I didn't have to look at changes that were already reviewed.

Reply:
Sorry about that, this must have made the review even harder...

> Nit: A better name would be _get_and_cache_course_override_value

IMHO it's not worth it to expose the internal cachine process in the method name. I'll keep this as it is, if you don't mind.